common/page.go

```diff
@@ -364,6 +363,127 @@ func (p *Page) initEvents() {
 	}()
 }
 
+func (p *Page) urlGroupingName(ctx context.Context, urlTag string) (string, bool) {
+	p.eventHandlersMu.RLock()
+
+	// If there are no handlers for EventConsoleAPICalled.
+	if _, ok := p.eventHandlers[EventPageMetricCalled]; !ok {
+		p.eventHandlersMu.RUnlock()
+
+		return "", false
+	}
+	p.eventHandlersMu.RUnlock()
+
+	var name string
+	var nameChanged bool
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	p.eventHandlersMu.RLock()
+
+	for _, h := range p.eventHandlers[EventPageMetricCalled] {
+		// A handler can register another handler from within itself. This is
+		// the reason to unlock the mutex before calling the handler.
+		p.eventHandlersMu.RUnlock()
+
+		nameCh := make(chan string)
+		em := &ExportedMetric{
+			ctx:    ctx,
+			urlTag: urlTag,
+			nameCh: nameCh,
+		}
+
+		// handler runs on another goroutine.
+		h(em)
```

Comment:
Where does it run in another goroutine? It seems like a blocking call here.

Reply:
So the handler is called, and within the handle there is a call to [add it to the taskqueue](https://github.com/grafana/xk6-browser/blob/fd8aa9778ca160ede9d775bc60c0799fabb49e97/browser/page_mapping.go#L233-L254). When something is ran on the task queue it is ran on a separate goroutine to what added the function to the task queue.

I was looking at the [waitForEvent](https://github.com/grafana/xk6-browser/blob/8fce16a75360a2317dfe2ff4c8ac4e5a2c292b5b/browser/browser_context_mapping.go#L113-L132) implementation and i'm going to attempt to work with that.

Reply:
I think i've made it a little easier now in ef24fb2782d1956292e2fa989fbf5e5b4e89f181. It doesn't rely on a channel to return the name updates, instead a new name filed is updated in ExportedMetric which can be read by the caller in `urlGroupingName`.

The call to `h(em)` is now also synchronous since we're synchronising the use of the taskqueue in the mapping layer, like with [waitForEvent](https://github.com/grafana/xk6-browser/blob/8fce16a75360a2317dfe2ff4c8ac4e5a2c292b5b/browser/browser_context_mapping.go#L113-L132).

Reply:
Perfect!